### PR TITLE
mesa can now be built on centos / rhel 6 servers

### PR DESCRIPTION
**Pull request by JusticeForMikeBrown**
@adamjstewart

### CODE REVIEW

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 1]
```diff
@@ -34,6 +34,8 @@ class Mesa(AutotoolsPackage):
 
     version('12.0.3', '60c5f9897ddc38b46f8144c7366e84ad')
 
+    variant('nodri', default=False, description='Disable DRI and EGL support')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: Can you make this positive? Name it `dri`, and say that it enables DRI and EGL support.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 2]
```diff
@@ -52,10 +54,28 @@ class Mesa(AutotoolsPackage):
...
+    # 
+    def configure_args(self):
+        spec = self.spec
+        args = []
+        if '+nodri' in spec:
+            args += ['--disable-dri']
+            args += ['--disable-egl']
+            args += ['--enable-sysfs']
+            args += ['--disable-driglx-direct']
+            args += ['--with-gallium-drivers=swrast']
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: You should also explicitly activate these for `+dri`.

Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: They're found and activated by default via the configure script.  So 

`variant('dri', default=Enable, description='Enable DRI and EGL support')`

Should be enough.

**Comment by JusticeForMikeBrown** [comment]
Let me know if I need to blow this away and submit new PR given the errant commits.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: It never hurts to be explicit. Also, it looks like you're creating 5 lists of length one and appending 5 times for a total of 10 lists, which isn't very efficient. Why not just use:
```python
args.extend([
    '--disable-dri',
    '--disable-egl',
    '--enable-sysfs',
    '--disable-driglx-direct',
    '--with-gallium-drivers=swrast',
])
```

**Comment by adamjstewart** [comment]
No need to submit a new PR, it's much better to keep everything in one PR so we don't have to scroll through multiple to see the comment history.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 2 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: @adamjstewart Thanks for the tip!  I'll add this now.

Comment (luigi-calori) [inline var/spack/repos/builtin/packages/mesa/package.py]: Sorry, as I'm also compiling mesa o headless cent'OS, see PR #4134 , coulld you have a look so 
either I test your pr on my side or I modify mine with your requirements?
For example, I' m using version 13 but got notified 17 is out.
also I' ve tested with good performance with gallium, and it seem to be needed the env var
'GALLIUM_DRIVER', 'swr'

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 3]
```diff
@@ -52,10 +55,43 @@ class Mesa(AutotoolsPackage):
...
+        args = []
+        if '~dri' in spec:
+            args.extend([
+                '--disable-dri',
+                '--disable-egl',
+                '--enable-sysfs',
+                '--disable-driglx-direct',
+                '--with-gallium-drivers=swrast'
+            ])
+        if '+dri' in spec:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: Just use `else:`

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 4]
```diff
@@ -52,10 +55,43 @@ class Mesa(AutotoolsPackage):
...
+            args.extend([
+                '--enable-dri',
+                '--enable-egl'
+            ])
+
+        if '+osmesa' in spec:
+            args.extend([
+                '--enable-osmesa',
+                '--disable-driglx-direct',
+                '--disable-dri',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: So if I build `mesa+dri+osmesa`, what do I get? You are enabling _and_ disabling dri...

**Comment by luigi-calori** [comment]
Sorry, I' m also building on headless centOS see #4134 , is it a similar use case?
I' m using version 13 and 17 is out and building with gallium swr, that needs an env var 
run_env.set('GALLIUM_DRIVER', 'swr')
could you have a look so I can try to follow your PR ?

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 4 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: You cannot do both.  This commit isn't doing both, either.  It's variant dri by default with osmesa disabled.

Demonstrating the mutual exclusivity:

`spack -d install mesa +dri+osmesa %gcc@6.3.0`

`--enable-dri' '--enable-egl' '--enable-osmesa' '--disable-driglx-direct' '--disable-dri' '--with-gallium-drivers=llvmpipe'
`
```
checking for PTHREAD_PRIO_INHERIT... no
checking for PTHREADSTUBS... yes
checking for LIBDRM... yes
checking for SHA1Init... no
checking for CC_SHA1_Init... no
checking for wincrypt.h... no
checking for SHA1Init in -lmd... no
checking for LIBSHA1... no
checking for nettle_sha1_init in -lnettle... no
checking for gcry_md_open in -lgcrypt... yes
checking for SHA1_Init in -lcrypto... yes
checking for OPENSSL... yes
checking for SHA1 implementation... libgcrypt
checking for LIBUDEV... no
checking for XLIBGL... yes
checking for mincore... yes
configure: error: egl requires --enable-dri
```

**Comment by JusticeForMikeBrown** [comment]
@luigi-calori 

I think we're trying to do the same thing here though I didn't add llvm.

Does yours build on Centos 6?  I couldn't get that to work without "--enable-sysfs."

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 4 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: If they're mutually exclusive, you need to make that explicit so it fails during concretization:
```python
conflicts('+dri', '+osmesa')
```
What about `mesa~dri~osmesa`? Does that work?

Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: BAM


```
user@node103[~/repos/spack]$ spack -d install mesa +dri+osmesa %gcc@6.3.0                                                                                                                                                                  
==> Reading config file /pbtech_mounts/homes027/user/repos/spack/etc/spack/defaults/packages.yaml
==> WRITE LOCK: /home/user/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquiring]
==> WRITE LOCK: /home/user/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Released]
==> Reading config file /home/user/.spack/compilers.yaml
==> Reading config file /home/user/.spack/linux/compilers.yaml
==> Error: Conflicts in concretized spec "mesa@12.0.3%gcc@6.3.0+dri+osmesa arch=linux-rhel6-x86_64 /xmw5u3s"

```

Yes, I already built it:

```
-- linux-rhel6-x86_64 / gcc@6.3.0 -------------------------------
r5fqjwo mesa@12.0.3%gcc~dri

lrcco52 mesa@12.0.3%gcc~dri~osmesa
```

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 1 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: This is done

**Comment by JusticeForMikeBrown** [comment]
@adamjstewart Thanks again for your help.  Please merge as soon as possible.  I have another R package to add, [rgl,](https://cran.r-project.org/web/packages/rgl/index.html) which was built successfully using mesa+dri and mesa~dri.

**Comment by adamjstewart** [comment]
@JusticeForMikeBrown Since there are multiple competing Mesa PRs open, I would like to hear back from others before merging, especially @luigi-calori.

**Comment by luigi-calori** [comment]
@JusticeForMikeBrown , @adamjstewart  In my PR, I tested 13.0.2 on compute nodes of our cluster, Marconi. I used gallium option and tested the installation with a binary version of ParaView.
As far as I understand, you need LD_LIBRARY_PATH defined to let binary OpenGL code to use Mesa
they are Broadwell nodes with  CentOS Linux release 7.2.1511
which version did you tested?

**Comment by luigi-calori** [comment]
@JusticeForMikeBrown It could be that verson 12 need "--enable-sysfs." on  CentOS 6 while version 13 need llvm or other combination.
I think that we should arrange a number of variants that allow both of us to build the version we want on our platform.
I added the changes that I found needed on my platform starting from develop version,   trying to be as less invasive as possible.
If you want to stick to older mesa version, ok I can add my depends on llvm just for  following version.
Regarding merge sequence, I would stick to @adamjstewart decision, if he think it would be better to merge your PR first, I' ll rebase mine accordingly.
I just want to move version up as I think that  ParaView support has improved after 12.
If I remember well, llvm is required for the high performance gallium driver.
As llvm is quite a big dep, I opted for a specific dependency
depends_on('llvm@3.9.1', when='@13.0.2+gallium')
If @adamjstewart think is better to merge mine first and something on my side is left, let me know, I' ll try to work on it this evening.

**Comment by JusticeForMikeBrown** [comment]
@luigi-calori 

"--enable-sysfs" has to do with the fact that Centos 6 has an older version of udev [147] while mesa 12 wants version 151 which can only be obtained in Centos 7.  

`I think that we should arrange a number of variants that allow both of us to build the version we want on our platform.`

That would be ideal since libudev should be chosen if we're dealing with Centos 7.  We are eventually retiring Centos 6 and do have Centos 7 servers as well.

**Comment by JusticeForMikeBrown** [comment]
[This](https://spack.readthedocs.io/en/latest/developer_guide.html) should work.  I'll add it to the package.

```
user@node103[~/repos/spack]$  spack python -c 'import distro; distro.linux_distribution()'
(u'RedHatEnterpriseServer', u'6.3', u'Santiago')
```

**Comment by JusticeForMikeBrown** [comment]
This does not work:

```
    def configure_args(self):
        spec = self.spec
        version = distro.linux_distribution()[1]
        # centos / rhel
        if version == range(6.0, 6.9):
            args = ['--enable-sysfs']
        else:
            args = []

```
```
=> Error: NameError: global name 'distro' is not defined
/pbtech_mounts/homes027/user/repos/spack/var/spack/repos/builtin/packages/mesa/package.py:77, in configure_args:
```

If I add "from spack import distro" I then get:

`cannot import name distro`

Do I rather need to `import os`?

This seems to be commonly used in other packages.

**Comment by adamjstewart** [comment]
Did you try `import distro`? Also, I don't think that code will work as you expect on Ubuntu 6 or Fedora 6 or any non-RHEL/CentOS system.

**Comment by JusticeForMikeBrown** [comment]
Fedora 6 and Ubuntu 6 are 10 years old.  I added this code so it would work on servers that are still supported and possibly in use though does it make sense to add conditionals to deal with situations are are highly unlikely to be encountered?

Thanks for all your help.  The latest commit now works with `import distro.`

`spack install mesa ~dri`

`'--enable-sysfs' '--disable-dri' '--disable-egl' '--enable-sysfs' '--disable-driglx-direct' '--with-gallium-drivers=swrast'`

**Comment by adamjstewart** [comment]
If you introduce code that you know only works on RHEL/CentOS 6, then you definitely need to check to make sure that you are on RHEL/CentOS 6. Checking that you are on some operating system at version 6 isn't good enough.

**Comment by JusticeForMikeBrown** [comment]
Ok.

I will add checks for `CentOS` and `RedHatEnterpriseServer`

**Comment by adamjstewart** [comment]
@JusticeForMikeBrown @luigi-calori @mjwoods 

We now have 3 competing PRs for the Mesa package that do similar things:
#4316, #4258, and #4134 

I personally don't know that much about Mesa (aside from it not building for me on CentOS 6). It would be great if you could all fight amongst yourselves and come to an agreement on which of these PRs is the best solution for everyone's problems and close the ones that you want to abandon.

**Comment by JusticeForMikeBrown** [comment]
@luigi-calori @mjwoods 

I don't care much either as long as it will build on centos / rhel 6 servers.

**Comment by JusticeForMikeBrown** [comment]
@luigi-calori @mjwoods 

Can we have this merged then you guys can change it as needed?

**Comment by luigi-calori** [comment]
On my side si ok, go ahead with the merge, I' ll rebase my stuff

**Comment by JusticeForMikeBrown** [comment]
@mjwoods 

What say you?

**Comment by JusticeForMikeBrown** [comment]
@adamjstewart 

Can we merge this now, please? @mjwoods has said nothing but @luigi-calori supports the decision.  The quicker it's merged the sooner others can make changes as they see fit.

**Comment by JusticeForMikeBrown** [comment]
@adamjstewart though, let me test the checks for `RedHatEnterpriseServer` and `CentOS` further.

**Comment by JusticeForMikeBrown** [comment]
` version = distro.linux_distribution()[1]` is not a string but unicode so this finally actually works:

```
        version = distro.linux_distribution()[1]
        distname = distro.linux_distribution()[0]
        # centos / rhel 6
        if distname == "RedHatEnterpriseServer":
            if version == (u'6.3' or u'6.4'):
                args = ['--enable-sysfs']
            else:
                args = []
        elif distname == "CentOS":
            if version == (u'6.3' or u'6.4'):
                args = ['--enable-sysfs']
            else:
                args = []
        else:
            args = []

```
When doing `spack install mesa ~dri` configure shows

`'--enable-sysfs' '--disable-dri' '--disable-egl' '--disable-driglx-direct' '--with-gallium-drivers=swrast'`

However, some regex would be nice rather than many `or` operators.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 5]
```diff
@@ -52,10 +57,59 @@ class Mesa(AutotoolsPackage):
...
+    # https://www.mesa3d.org/osmesa.html
+    # https://www.mesa3d.org/autoconf.html
+    def configure_args(self):
+        spec = self.spec
+        version = distro.linux_distribution()[1]
+        distname = distro.linux_distribution()[0]
+        strversion = version.encode('utf-8')
+        # centos / rhel 6
+        if distname == "RedHatEnterpriseServer":
+            if strversion.find('6.0:6.9') == -1:
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: I don't think this is how `find` works. The substring `6.0:6.9` will never be found. You should convert the value to a `Version()` object and see if the version is what you expect:
```python
if Version('6.0') <= Version(version) <= Version('6.9'):
```

Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: Thanks!

Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: @adamjstewart 

This seems to work

```
>>> print(version)
6.3
>>> type(version)
<type 'unicode'>
>>> if '6.0' <= str(version) <= '6.9':
...     print("true")
... else:
...     print("false")
... 
true
```

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: It does, some of the time:
```python
>>> '6.0' <= '6.5' <= '6.10'
False
```
Are you sure there will never be a 6.10 release?

Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: Ha ha!  @adamjstewart 

No, and actually Red Hat I believe did use that numbering before Enterprise Linux became a thing.

Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: There was a RHEL 5.10

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: Seeing as how there was a RHEL 5.11 release, you should probably use `Version()` and expand the range to something greater:
```
$ spack python
>>> from spack.version import *
>>> Version('6.0') <= Version('6.11') <= Version('6.999')
True
```

**Comment by mjwoods** [comment]
My apologies for the delay in replying. I need more time to examine the different proposals before I can offer an informed opinion.

To handle the selection of libsysfs vs libudev, I am exploring a method that uses trial and error during the configure stage. This would be more flexible (and probably more reliable) than basing the choice on platform labels. I will upload some example code on the weekend.

I understand that people want a solution quickly, but Mesa has a lot of options, and we want to support a range of platforms. Let's try to get this package right.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 5 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: @adamjstewart 

Got you.

```
Spack version 0.10.0
Python 2.7.13, Linux x86_64
>>> from spack.version import *
>>> import distro
>>> Version('6.0') <= Version(distro.linux_distribution()[1]) <= Version('6.999')
True
>>> Version(distro.linux_distribution()[1])
Version('6.3')
```

**Comment by JusticeForMikeBrown** [comment]
@mjwoods 

Right now Mesa, as it exists within Spack, is broken for RHEL / CentOS 6.  I went though the trial and error.  It's impossible to use libudev as the needed version cannot be installed.  You must use libsys.  Given that I think for this distro `Version()` would be the appropriate choice.  Your solution might be good for other distributions though.

**Comment by JusticeForMikeBrown** [comment]
Some notes.

12.0.3 builds with software rendering `~dri` though both 12.0.3 and 17.1.1 fail for `+dri`.

12.0.3 fails due to cmake issue with `llvm@3:3.99`

17.1.1 due to needing `libdrm >= 2.4.75 `which is absent from Spack.

Right now I don't care much about hardware rendering though I do want to add VMD to Spack, which [lists](https://www-s.ks.uiuc.edu/Research/vmd/doxygen/compiling.html#compiling) OpenGL as a requirement.  So it will be needed.

**Comment by JusticeForMikeBrown** [comment]
## Addendum

17.0.1 ~dri now works.  Perhaps I forgot a comma.

`spack install mesa~dri %gcc@6.3.0`

```
==> Already patched mesa
==> Building mesa [AutotoolsPackage]
==> Executing phase : 'autoreconf'
==> Executing phase : 'configure'
==> Executing phase : 'build'
==> Executing phase : 'install'
==> Successfully installed mesa
  Fetch: 0.02s.  Build: 2m 8.08s.  Total: 2m 8.10s.
```

Perhaps a variant for gdm should be added.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 6]
```diff
@@ -23,16 +23,26 @@
 # Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 ##############################################################################
 from spack import *
+from spack.version import *
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: This is already included in `from spack import *`.

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 7]
```diff
@@ -23,16 +23,26 @@
...
 
     homepage = "http://www.mesa3d.org"
-    url      = "http://ftp.iij.ad.jp/pub/X11/x.org/pub/mesa/12.0.3/mesa-12.0.3.tar.gz"
+    url      = "https://mesa.freedesktop.org/archive/12.0.3/mesa-12.0.3.tar.xz"
 
-    version('12.0.3', '60c5f9897ddc38b46f8144c7366e84ad')
+    list_url = "https://mesa.freedesktop.org/archive"
+
+    version('12.0.3', '1113699c714042d8c4df4766be8c57d8')
+    version('17.1.1', 'a4844bc6052578574f9629458bcbb749')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: Can you put newer versions above older versions?

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 8]
```diff
@@ -52,10 +62,64 @@ class Mesa(AutotoolsPackage):
...
+        if distname == "RedHatEnterpriseServer":
+            if Version('6.0') <= Version(v) <= Version('6.999'):
+                args = ['--enable-sysfs']
+            else:
+                args = []
+        elif distname == "CentOS":
+            if Version('6.0') <= Version(v) <= Version('6.999'):
+                args = ['--enable-sysfs']
+            else:
+                args = ['--enable-sysfs']
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: Both of these are enable. Did you mean disable?

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 9]
```diff
@@ -52,10 +62,64 @@ class Mesa(AutotoolsPackage):
...
+        v = distro.linux_distribution()[1]
+        distname = distro.linux_distribution()[0]
+
+        # libudev too old on centos / rhel 6 so do sysfs
+        if distname == "RedHatEnterpriseServer":
+            if Version('6.0') <= Version(v) <= Version('6.999'):
+                args = ['--enable-sysfs']
+            else:
+                args = []
+        elif distname == "CentOS":
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: Why not do `if distname == RHEL or disname == CentOS`? Then you don't need to repeat all of this.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 8 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: @adamjstewart fixed!

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 9 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: Great idea!

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 7 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: Makes sense.

File: var/spack/repos/builtin/packages/mesa/package.py (hunk 6 above)
Comment (JusticeForMikeBrown) [inline var/spack/repos/builtin/packages/mesa/package.py]: Removed

**Comment by JusticeForMikeBrown** [comment]
Mesa 17.1.1 with Hardware Rendering `+dri` now compiled successfully using `llvm@4` with variant `+link_dylib`

```

==> Ran patch() for llvm
==> Building llvm [CMakePackage]
==> Executing phase : 'cmake'
==> Executing phase : 'build'
==> Executing phase : 'install'
==> Successfully installed llvm
  Fetch: 0.32s.  Build: 18m 10.59s.  Total: 18m 10.91s.
  [+] /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/llvm-4.0.0-2fimg4y6mvngu4ileaepor3qfof5vmfv
  ==> bison is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/bison-3.0.4-f34eyvtqhh4jr7svw2bfsh375ldtvmdq
  ==> dri2proto is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/dri2proto-2.8-af2r3jhrmp7xtlw45m4nkvh43jbkm5gk
  ==> python is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/python-2.7.13-hccdc77sssvp6xnoicrpnub4ujen7ehe
  ==> glproto is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/glproto-1.4.17-cm55nxwtvavbhfro64moeehdiy3ha27u
  ==> libpthread-stubs is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/libpthread-stubs-0.4-ess3yk6ivtx5mzxssq5nwyafbbcz657s
  ==> libxdamage is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/libxdamage-1.1.4-4b553vfkrdf5yotptyz4ksomk6fo3t6h
  ==> pkg-config is already installed in /pbtech_mounts/homes027/user/repos/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/pkg-config-0.29.2-siiecvsigui7s5vdwne6rhsjg36ywtzs
  ==> Installing mesa
  ==> Using cached archive: /pbtech_mounts/homes027/user/repos/spack/var/spack/cache/mesa/mesa-17.1.1.tar.xz
  ==> Staging archive: /pbtech_mounts/homes027/user/repos/spack/var/spack/stage/mesa-17.1.1-hjw5hcykccoxinwafu6gsv74ya6oghcg/mesa-17.1.1.tar.xz
  ==> Created stage in /pbtech_mounts/homes027/user/repos/spack/var/spack/stage/mesa-17.1.1-hjw5hcykccoxinwafu6gsv74ya6oghcg
  ==> Ran patch() for mesa
  ==> Building mesa [AutotoolsPackage]
  ==> Executing phase : 'autoreconf'
  ==> Executing phase : 'configure'
  ==> Executing phase : 'build'
==> Executing phase : 'install'
==> Successfully installed mesa
  Fetch: 0.02s.  Build: 5m 45.70s.  Total: 5m 45.73s.
```

**Comment by JusticeForMikeBrown** [comment]
I added variant `+gdm` to compile Mesa with GDM support.  Though default variant `~gdm` given we're likely headless.

**Comment by mjwoods** [comment]
Hi @JusticeForMikeBrown , @luigi-calori and @adamjstewart .

For what it's worth, I've updated #4258 to include automatic selection of libudev or libsysfs for versions of mesa that need them. I've also tried to keep the variants to a minimum to limit the number of cases that need testing.

I have managed to build three "safe" versions of mesa (12.0.3, 13.0.6 and 17.1.1) on RHEL6 using different compilers ranging from `gcc@4.4.7` to `gcc@6.3.0`. Only the variant `~dri~gallium` could be built with `gcc@4.4.7` due to lack of support for the `c++11` standard. Newer compilers could build more of the variants.

Mesa provides a large number of options and supports many types of hardware, so it is difficult to create a simple spack package that meets every need. I have taken the approach of building a comprehensive range of features that will suit most applications, even though the compilation time is increased by building features that may not be used. I suspect that the most common variants will be `~dri~gallium` and `~dri+gallium`, which will be useful on the kinds of servers used for scientific computing. People with their own workstations have the option of installing a hardware accelerated mesa via the system package manager.

I ask that you review #4258 and either merge it into `develop` or cherry-pick the parts you want for your branch.

**Comment by davydden** [comment]
@JusticeForMikeBrown is this PR still active? Maybe rebase on top of develop?

**Comment by alalazo** [comment]
@JusticeForMikeBrown I am closing this as inactive. Feel free to reopen if there are updates

**Pull request closed without merging**